### ISSUE TITLE
Date is not removed from book history list when all items from that date are removed.

### ISSUE THREAD

**Issue by Frans-Lukas** [kiwix/kiwix-android#1990]
**Describe the bug**

When all history items from a specific date are deleted, the date stamp remains. This looks weird and is not the expected behavior. 

**Expected behavior**

If all history items within a specific date are deleted, the date stamp should also be deleted from the list. 

**Steps to reproduce the behavior:**
1. Have a list of history items from multiple dates.
2. Open the history activity.
3. Delete all items from a specific date.
4. See that the timestamps remain. 

**Screenshots**

![dateDoesNotDisappear](https://user-images.githubusercontent.com/22193232/78029603-2a7bcd80-7361-11ea-8c3b-ba33d8ead534.gif)

I'd like to work on this.

**Comment by macgills** [comment]
@Frans-Lukas if you are looking for a challenge these screens(history and bookmarks which are essentially a copy paste) are not adhering to Unidirectional Data Flow. It would not be an inconsiderate rewrite but these screens have a host of data flow issues because of their outdated architecture

**Comment by Frans-Lukas** [comment]
I might be looking for one! I'll take a look and see if I get what to do.

**Comment by macgills** [comment]
You can look at SearchActivity or LanguageActivity for some simple examples.

The basic principle is the View sends Actions to the ViewModel, the ViewModel processes the Actions and emits States, the View renders States.

It can be quite challenging because you have to explicitly model the entire state of the screen. It will also require some knowledge of Rx and reactive data flows in general. I am here to answer questions

**Comment by Frans-Lukas** [comment]
Ok, I haven't done any Rx so I'll have to look into that and make sure to ask questions if they arise.

**Comment by macgills** [comment]
We could then have the adapter use the AdapterDelegates pattern and it would be much nicer

**Comment by Frans-Lukas** [comment]
@macgills How does this work? 

```
private fun viewStateReducer() =
    Flowable.combineLatest(
      recentSearchDao.recentSearches(zimReaderContainer.id),
      searchResultsFromZimReader(),
      filter,
      searchOrigin,
      Function4(this::reduce)
    ).subscribe(state::postValue, Throwable::printStackTrace)
```

My thinking is that it 'keeps track' of each of the search result-publisher and combines them when any of them update to allow the state to change. But what is Function4/a combiner?

**Comment by macgills** [comment]
In english "whenever any of these data sources emits something take that new value and the latest known value of all the other data sources and use the function `reduce` to create a new state object" and then the subscribe takes the value and sends it to the livedata `state`

In strict redux a reducer is only `previousState + newAction = newState` so I was cheating a little bit but this screen is not very complex

**Comment by Frans-Lukas** [comment]
ok, so for the historyViewModel it might use the historyDao, the filter and the reducer? Ah, originally I thought all emitters emitted the same type of value but I see now that they are emitting different types of values that are used in combination to get the correct search results. Cool!

**Comment by macgills** [comment]
It would also use actions `ShowAllSwitchToggled` `UserClickedItem`, LanguageViewModel shows an example of this

**Comment by Frans-Lukas** [comment]
Ok, cool. Thanks!

**Comment by Frans-Lukas** [comment]
@macgills Should I create a separate issue for the HistoryActivity rewrite? 

I have some WIP code that I'd like to have reviewed, do I just do a normal pull request with WIP in the title for that?

**Comment by macgills** [comment]
We can use this ticket as the excuse for a rewrite. Open a "draft" PR and I can give you feedback

**Comment by Frans-Lukas** [comment]
Fixed by #2124 and #2119 .
